Off-Top-App/SessionReport
# OT-SessionReport | Created a new microservice that outputs a report of a user's session

**What I Did:**
- I created a new microservice that makes a report of a user's session.
- I used Kafka Streams to consume, process, and produce data.

**How to Test:**
- Have Kafka and Zookeeper open.
- Install [Kotlin and IntelliJ](https://kotlinlang.org/docs/tutorials/jvm-get-started.html). Unfortunately Kotlin doesn't work well with other IDEs. Fortunately, all you need to install is IntelliJ and you'll get Kotlin with it.
- Run the new microservice.
- Open a terminal and run the following Kafka command: 
-> Mac/Docker: `kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic NewSessionReport`.
-> Windows: `kafka-console-consumer.bat --bootstrap-server localhost:9092 --topic NewSessionReport`.
- For ease of testing, [I've provided a testing script using Python](https://drive.google.com/file/d/1_NGldDftA05iJAwHyxQCW_roS5Dx-cOA/view?usp=sharing). You are free to create your own if you like.
>Incoming data should look like this: 
```
{
    user_id: int, 
    session_data: [
        {   
            focus_score: boolean,
            focus_value: int,
            analyzed_at: timestamp (String),
            time_exported: timestamp (String),
            transcribed_speech: string,
            topic: string
        },
        {
        ...
        },
        {   
            focus_score: boolean,
            focus_value: int,
            analyzed_at: timestamp (String),
            time_exported: timestamp (String),
            transcribed_speech: string,
            topic: string
        }
    ],
}
```
>Outgoing data will look like this:
```
{
    user_id: int, 
    session_data: [
        {   
            focus_score: boolean,
            focus_value: int,
            analyzed_at: timestamp (String),
            time_exported: timestamp (String),
            transcribed_speech: string,
            topic: string
        },
        {
        ...
        },
        {   
            focus_score: boolean,
            focus_value: int,
            analyzed_at: timestamp (String),
            time_exported: timestamp (String),
            transcribed_speech: string,
            topic: string
        }
    ],
    average_focus_score: double
    topic: string,
    most_used_word: [string, string, string],
    start_time: timestamp (String),
    end_time: timestamp (String),
    session_duration: timestamp (String)
}
```
- Run the Python script. You should see in the terminal for Kafka that a new message has produced for the topic `NewSessionReport` containing the new data. 
![image](https://user-images.githubusercontent.com/36316432/88509028-050c7280-cf95-11ea-9917-f5d1436a276c.png)
- Open a new terminal and run the following command: 
->Mac/Docker: `kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic NewUserReport`.
->Windows: `kafka-console-consumer.sh --bootstrap-server localhost:9092 --topic NewUserReport`.
- Again, run the same Python script and you should get the following message to a different topic, `NewUserReport`:
![image](https://user-images.githubusercontent.com/36316432/88509088-2f5e3000-cf95-11ea-8f4f-4742c6762e36.png)
- If the data has been successfully produced in the two topics, then everything is good!
